rasgotransforms/rasgotransforms/jinja.py

```diff
@@ -0,0 +1,60 @@
+from jinja2 import Environment
+from typing import Callable, Dict, Optional, Tuple, Union
+from itertools import combinations, permutations, product
+import re
+
+JINJA_ENV = Environment(extensions=['jinja2.ext.do', 'jinja2.ext.loopcontrols'])
+
+
+def cleanse_template_symbol(symbol: str) -> str:
+    symbol = str(symbol)
+    symbol = symbol.strip()
+    symbol = symbol.replace(' ', '_').replace('-', '_')
+    symbol = symbol.upper()
+    symbol = re.sub('[^A-Z0-9_]+', '', symbol)
+    symbol = '_' + symbol if symbol[0].isdecimal() or not symbol else symbol
+
+    return symbol
+
+
+def raise_exception(message: str) -> None:
+    raise Exception(message)
+
+
+def render(
+    source_code: str,
+    source_table: str,
+    arguments: dict,
+    source_columns: Optional[Union[Dict[str, Dict[str, str]], Callable]] = None,
+    run_query: Optional[Callable] = None,
+) -> str:
```

Comment:
Instead of expecting discrete function overrides, why not accept a dict of overrides that we can merge onto the local `template_fns` that way clients can choose which methods to overwrite (or add their own where necessary?)

Reply:
+1

Reply:
Should I leave `run_query` as it's own arg or just take it in the override dict? My thought was that if run_query was provided, any other macros could be defined in the jinja templates themselves. The more macros we allow to live outside of the jinja template, the more we require our clients to implement and pass in. 

If the only macro that we externally define here is `run_query` and force the transforms to implement any other macros inline with jinja, then it makes it much easier for clients to integrate. They only need to implement `run_query` which should be the only client specific macro, defining how they execute sql against the DW.